File: 2360-longest-cycle-in-a-graph/2360-longest-cycle-in-a-graph.py
class Solution:
    def longestCycle(self, edges: List[int]) -> int:
        n = len(edges)
        deg = [0 for _ in range(n)]
        adj = defaultdict(list)
        
        for u, v in enumerate(edges):
            if v != -1:
                deg[v] += 1
                adj[u].append(v)
        
        dq = deque()
        color = [0 for _ in range(n)]
        
        for u in range(n):
            if deg[u] == 0:
                dq.append(u)
                color[u] = 2
                
        while dq:
            u = dq.popleft()
            
            for v in adj[u]:
                if color[v] == 2:
                    continue
                deg[v] -= 1
                if deg[v] == 0:
                    color[v] = 2
                    dq.append(v)
                    
        def dfs(u):
            color[u] = 1
            cnt = 1
            for v in adj[u]:
                if color[v] == 0:
                    cnt += dfs(v)
                elif color[v] == 1:
                    break 
            color[u] = 2
            return cnt
        
        ans = -1
        
        for u in range(n):
            if color[u] != 0:
                continue
            ans = max(ans, dfs(u))
        
        return ans

File: 2360-longest-cycle-in-a-graph/2360-longest-cycle-in-a-graph.cpp
class Solution {
private:
    vector<int> degree, visited;
    
public:
    int longestCycle(vector<int>& edges) {
        const int n = edges.size();
        
        degree.resize(n);
        visited.resize(n);
        
        for (int i = 0; i < n; ++i) {
            if (edges[i] != -1) {
                degree[edges[i]]++;
            }
        }
        
        queue<int> mq;
        
        for (int i = 0; i < n; ++i) {
            if (degree[i] == 0) {
                mq.push(i);
            }
        }
        
        while (!mq.empty()) {
            int u = mq.front();
            mq.pop();
            visited[u] = true;
            
            int v = edges[u];
            
            if (v == -1) {
                continue;
            }
            
            degree[v]--;
            if (degree[v] == 0) {
                mq.push(v);
            }
        }
        
        int ans = -1;
        
        for (int i = 0; i < n; ++i) {
            if (!visited[i]) {
                int u = i;
                int cnt = 0;
                
                while (!visited[u]) {
                    cnt++;
                    visited[u] = true;
                    u = edges[u];
                }
                
                ans = max(ans, cnt);
            }
        }
        
        return ans;
    }
};